services/httpd/pprof.go

```diff
@@ -0,0 +1,151 @@
+package httpd
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"net/http"
+	httppprof "net/http/pprof"
+	"runtime/pprof"
+	"strconv"
+	"time"
+)
+
+// handleProfiles determines which profile to return to the requester.
+func handleProfiles(w http.ResponseWriter, r *http.Request) {
+	switch r.URL.Path {
+	case "/debug/pprof/cmdline":
+		httppprof.Cmdline(w, r)
+	case "/debug/pprof/profile":
+		httppprof.Profile(w, r)
+	case "/debug/pprof/symbol":
+		httppprof.Symbol(w, r)
+	case "/debug/pprof/all":
+		archiveProfiles(w, r)
+	default:
+		httppprof.Index(w, r)
+	}
+}
+
+// prof describes a profile name and a debug value, or in the case of a CPU
+// profile, the number of seconds to collect the profile for.
+type prof struct {
+	Name  string
+	Debug int64
+}
+
+// archiveProfiles collects the following profiles:
+//	- goroutine profile
+//	- heap profile
+//	- blocking profile
+//	- (optionally) CPU profile
+//
+// All profiles are added to a tar archive and then compressed, before being
+// returned to the requester as an archive file.  To optionally include a CPU
+// profile, the requester should provide a `cpu` query parameter, and can also
+// provide a `seconds` parameter to specify a non-default profile collection
+// time. The default CPU profile collection time is 30 seconds.
+//
+// Example request including CPU profile:
+//
+//	http://localhost:8086/debug/pprof/all?cpu=true&seconds=45
+//
+// The value after the `cpu` query parameter is not actually important, as long
+// as there is something there.
+//
+func archiveProfiles(w http.ResponseWriter, r *http.Request) {
+	var all = []*prof{
+		{Name: "goroutine", Debug: 1},
```

Comment:
@mark-rushakoff I think you misunderstood what I meant. `debug=2` is not compatible with `go tool pprof`, though it is of course valid with the `runtime/pprof` package handler, and will output something useful (to a human). It's just you can't feed that profile into the `pprof` tool.

```
edd@Alameda:~|
⇒  go tool pprof $GOPATH/bin/ossinfluxd-v1.2.1 http://localhost:8086/debug/pprof/goroutine\?debug\=1
Fetching profile from http://localhost:8086/debug/pprof/goroutine?debug=1
Saved profile in /Users/edd/pprof/pprof.ossinfluxd-v1.2.1.localhost:8086.goroutine.001.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) quit
edd@Alameda:~|
⇒  go tool pprof $GOPATH/bin/ossinfluxd-v1.2.1 http://localhost:8086/debug/pprof/goroutine\?debug\=2
Fetching profile from http://localhost:8086/debug/pprof/goroutine?debug=2
parsing profile: unrecognized profile format
edd@Alameda:~|
⇒  go tool pprof $GOPATH/bin/ossinfluxd-v1.2.1 http://localhost:8086/debug/pprof/goroutine\?debug\=0
Fetching profile from http://localhost:8086/debug/pprof/goroutine?debug=0
Saved profile in /Users/edd/pprof/pprof.ossinfluxd-v1.2.1.localhost:8086.goroutine.002.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) quit
edd@Alameda:~|
⇒
```